storage/store_test.go

```diff
@@ -1003,62 +1047,181 @@ func TestStoreResolveWriteIntentNoTxn(t *testing.T) {
-	}
+	for _, canPush := range []bool{true, false} {
+		keyA := proto.Key(fmt.Sprintf("%t-a", canPush))
+		keyB := proto.Key(fmt.Sprintf("%t-b", canPush))
 
-	// Next, write intents for keyA and keyB.
-	args.Value.Bytes = []byte("value2")
-	txnA := newTransaction("test", keyA, 1, proto.SERIALIZABLE, store.ctx.Clock)
-	txnB := newTransaction("test", keyB, 1, proto.SERIALIZABLE, store.ctx.Clock)
-	for _, txn := range []*proto.Transaction{txnA, txnB} {
-		args.Key = txn.Key
-		args.Timestamp = txn.Timestamp
-		args.Txn = txn
+		// First, write keyA.
+		args, reply := putArgs(keyA, []byte("value1"), 1, store.StoreID())
 		if err := store.ExecuteCmd(args, reply); err != nil {
 			t.Fatal(err)
 		}
-	}
 
-	// Now, get from both keys and verify.
-	gArgs, gReply := getArgs(keyA, 1, store.StoreID())
-	gArgs.ReadConsistency = proto.INCONSISTENT
-	if err := store.ExecuteCmd(gArgs, gReply); err != nil {
-		t.Errorf("expected read to succeed: %s", err)
-	} else if gReply.Value == nil || !bytes.Equal(gReply.Value.Bytes, []byte("value1")) {
-		t.Errorf("expected value %q, got %+v", []byte("value1"), gReply.Value)
-	}
-	gArgs.Key = keyB
-	if err := store.ExecuteCmd(gArgs, gReply); err != nil {
-		t.Errorf("expected read to succeed: %s", err)
-	} else if gReply.Value != nil {
-		t.Errorf("expected value to be nil, got %+v", gReply.Value)
+		// Next, write intents for keyA and keyB. Note that the
+		// transactions have unpushable priorities if canPush is true and
+		// very pushable ones otherwise.
+		priority := int32(-proto.MaxPriority)
+		if canPush {
+			priority = -1
+		}
+		args.Value.Bytes = []byte("value2")
+		txnA := newTransaction("testA", keyA, priority, proto.SERIALIZABLE, store.ctx.Clock)
+		txnB := newTransaction("testB", keyB, priority, proto.SERIALIZABLE, store.ctx.Clock)
+		for _, txn := range []*proto.Transaction{txnA, txnB} {
+			args.Key = txn.Key
+			args.Timestamp = txn.Timestamp
+			args.Txn = txn
+			if err := store.ExecuteCmd(args, reply); err != nil {
+				t.Fatal(err)
+			}
+		}
+		// End txn B, but without resolving the intent.
+		etArgs, etReply := endTxnArgs(txnB, true, 1, store.StoreID())
+		etArgs.Timestamp = txnB.Timestamp
+		if err := store.ExecuteCmd(etArgs, etReply); err != nil {
+			t.Fatal(err)
+		}
+
+		// Now, get from both keys and verify. Wether we can push or not,
+		// we will be able to read with INCONSISTENT.
+		gArgs, gReply := getArgs(keyA, 1, store.StoreID())
+		gArgs.Timestamp = store.ctx.Clock.Now()
+		gArgs.ReadConsistency = proto.INCONSISTENT
+		if err := store.ExecuteCmd(gArgs, gReply); err != nil {
+			t.Errorf("expected read to succeed: %s", err)
+		} else if gReply.Value == nil || !bytes.Equal(gReply.Value.Bytes, []byte("value1")) {
+			t.Errorf("expected value %q, got %+v", []byte("value1"), gReply.Value)
+		}
+		gArgs.Key = keyB
+		if err := store.ExecuteCmd(gArgs, gReply); err != nil {
+			t.Errorf("expected read to succeed: %s", err)
+		}
+		// The new value of B will be read as we can push txn B.
```

Comment:
this should also resolve the intent on B (though asynchronously), right? Are you checking that?